src/Client/HedgedConnections.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+#if defined(OS_LINUX)
+
+#include <Client/GetHedgedConnections.h>
+#include <Client/IConnections.h>
+#include <functional>
+#include <queue>
+
+namespace DB
+{
+
+class HedgedConnections : public IConnections
+{
+public:
+    using ReplicaStatePtr = GetHedgedConnections::ReplicaStatePtr;
+
+    HedgedConnections(const ConnectionPoolWithFailoverPtr & pool_,
+                      const Settings & settings_,
+                      const ConnectionTimeouts & timeouts_,
+                      const ThrottlerPtr & throttler,
+                      PoolMode pool_mode,
+                      std::shared_ptr<QualifiedTableName> table_to_check_ = nullptr);
+
+    void sendScalarsData(Scalars & data) override;
+
+    void sendExternalTablesData(std::vector<ExternalTablesData> & data) override;
+
+    void sendQuery(
+        const ConnectionTimeouts & timeouts,
+        const String & query,
+        const String & query_id,
+        UInt64 stage,
+        const ClientInfo & client_info,
+        bool with_pending_data) override;
+
+    Packet receivePacket() override;
+
+    Packet receivePacketUnlocked(AsyncCallback async_callback) override;
+
+    void disconnect() override;
+
+    void sendCancel() override;
+
+    Packet drain() override;
+
+    std::string dumpAddresses() const override;
+
+    size_t size() const override { return replicas.size(); }
+
+    bool hasActiveConnections() const override { return !active_connections_count_by_offset.empty(); }
+
+private:
+    /// We will save actions with replicas in pipeline to perform them on the new replicas.
+    class Pipeline
+    {
+    public:
+        void add(std::function<void(ReplicaStatePtr &)> send_function);
+
+        void run(ReplicaStatePtr & replica);
+    private:
+        std::vector<std::function<void(ReplicaStatePtr &)>> pipeline;
+    };
+
+    Packet receivePacketFromReplica(ReplicaStatePtr & replica, AsyncCallback async_callback = {});
+
+    Packet receivePacketImpl(AsyncCallback async_callback = {});
+
+    void processReceiveData(ReplicaStatePtr & replica);
+
+    void processTimeoutEvent(ReplicaStatePtr & replica, TimerDescriptorPtr timeout_descriptor);
+
+    void tryGetNewReplica();
+
+    void finishProcessReplica(ReplicaStatePtr & replica, bool disconnect);
+
+    int getReadyFileDescriptor(AsyncCallback async_callback = {});
+
+    GetHedgedConnections get_hedged_connections;
+
+    /// All replicas in replicas[offset] are responsible for process query
+    /// with setting parallel_replica_offset = offset. In common situations
+    /// replicas[offset].size() = 1 (like in MultiplexedConnections).
+    std::vector<std::vector<ReplicaStatePtr>> replicas;
```

Comment:
I think it would be logical to add another one structure like `OffsetState` with a list of replicas for offset, active connections count and so on.